server/src/main/java/org/elasticsearch/plugins/PluginsService.java

```diff
@@ -516,10 +525,20 @@ private Plugin loadBundle(Bundle bundle, Map<String, Plugin> loaded) {
 
         Class<? extends Plugin> pluginClass = loadPluginClass(bundle.plugin.getClassname(), loader);
         Plugin plugin = loadPlugin(pluginClass, settings, configPath);
+        // need to do this transitively - to some extent because of the jarhell check, to be discussed.
+        transitivePlugins(bundle.plugin.getExtendedPlugins(), lookup).map(loaded::get).map(ExtensiblePlugin.class::cast)
```

Comment:
Calling the `extensionPlugin` method with every extending plugin - transitively to support cases like sql -> ql -> x-pack-core.

Reply:
That transitive case should already be handed by the sortBundles method. Can you describe more why that doesn't work, what this is doing differently, and why we would need both?

Reply:
I believe `sortBundles` only sorts the bundles. The case I ran into was the one described in the comment, where we have following dependencies:

sql -> ql -> x-pack-core

sql cannot declare to also depend directly on x-pack-core, due to the jarhell check. If we only added it to those plugins that it extends directly, it would not be added to the `XPackPlugin` (only to the `QLPlugin`). By transitively searching for plugins, we add it to all direct and indirect dependencies.

I think the same applies to the `reloadSPI` part above.

Reply:
That sounds like a bug in sort then. The "sorting" is supposed to come up with the correct loading order, by inspecting extended plugins. IIRC, the fact any xpack plugins depend on core is a hack because they require XPackLicenseState, which is grabbed statically, so they assume xpack core has already been created, and can pull it. sql does depend on x-pack-core, as a compileOnly dependency, for this reason. The fact sql does not include x-pack-core in extendedPlugins is a bug, and doing so causing jar hell is also a bug.

Reply:
The "sorting" does generate the loading order (as in innner-most plugin first). But it does not help regarding transitiveness.

Are you saying that an x-pack component can use any class/interface from x-pack-core without declaring a dependency on it? I think most x-pack components use some x-pack-core class and I think the way we build the class-loaders require an explicit `extendedPlugin`. But the class-loaders themselves work transitively AFAICS.

Ack on the transitive vs. explicit dependency. I prefer explicit too. I think ideally, the classloaders should then be setup accordingly, but let us leave that for a future exercise. Once we settle on the plugin discussion I can work on removing the transitiveness here and fix the jar hell check. Provided that code survives that discussion :slightly_smiling_face: .